### DISCUSSION TITLE
Error when using with Cygwin

### DISCUSSION THREAD

**Discussion by RT-Tap** [JanDeDobbeleer/oh-my-posh#986] [Troubleshoot]
Hey, first off I want to thank all contributors for this really cool project!  Also I know Cygwin isn't officially supported however it seems to work or at least the base functionality does.

So I successfully got oh-my-posh working without much effort or changing anything just used the linux version of oh-my-posh on Cygwin.  However I keep getting an error and  the path portion of a theme says `INVALID CONFIG PATH`.  The Error is as follows:
```
Oh My Posh Error:
 CreateFile /cygdrive/c/Users/me/iterm2.omp.json: The system cannot find the path specified.
```
I believe this is because Cygwin prefaces normal system directories with `/cygwin/` so for example `C:/Users/Me/Documents/` would be  `/cygwin/c/Users/Me/Documents/` and directories in cygwin's directory aren't changed at all, or rather the change is it treats it's root directory as the root directory and anything outside is prefaced with `/cygwin/`.  However I get this error even when in Cygwin's home directory where using the command `pwd` would result in a path like `/home/me`.

Any Idea how i could fix this?  Would I need to compile from scratch and change some file location configs or something along those lines?  

Maybe I can do some enviroment variable changes?  However I don't want to break literally everything else so doubt this is a good idea.

Also to save HD space (minimal I know but efficiency) I simply symlinked `.poshthemes` in my cygwin directory to the themes folder in my Windows instance but I don't think this is part of the issue or at least doesnt seem to be.

![ohmyposherror](https://user-images.githubusercontent.com/71105781/134434623-1d39d605-d3e3-41dd-9756-079b0147691e.png)

**Comment by JanDeDobbeleer** [comment]
What's your bash config like?

**Reply by RT-Tap** [reply]
Sorry if I misinterpreted you question or just ignorant to a different config but in `.bashrc` I only have the following:
```
[[ "$-" != *i* ]] && return
eval "$(oh-my-posh --init --shell bash --config ~/iterm2.omp.json)"
```

EDIT:  in `/cygwin/etc/bash.bashrc` I have
```
# Check that we haven't already been sourced.
[[ -z ${CYG_SYS_BASHRC} ]] && CYG_SYS_BASHRC="1" || return

# If not running interactively, don't do anything
[[ "$-" != *i* ]] && return

# Exclude *dlls from TAB expansion
export EXECIGNORE="*.dll"

# Set a default prompt of: user@host and current_directory
PS1='\[\e]0;\w\a\]\n\[\e[32m\]\u@\h \[\e[33m\]\w\[\e[0m\]\n\$ '
```
In  `/cygwin/etc/profile` i have:

```
# -*- mode:sh sh-shell:sh -*-
# To the extent possible under law, the author(s) have dedicated all
# copyright and related and neighboring rights to this software to the
# public domain worldwide. This software is distributed without any warranty.
# You should have received a copy of the CC0 Public Domain Dedication along
# with this software.
# If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

# base-files version 4.3-3

# /etc/profile: executed by the command interpreter for login shells.

# The latest version as installed by the Cygwin Setup program can
# always be found at /etc/defaults/etc/profile

# Modifying /etc/profile directly will prevent
# setup from updating it.

# System-wide profile file

# Some resources...
# Customizing Your Shell: http://www.dsl.org/cookbook/cookbook_5.html#SEC69
# Consistent BackSpace and Delete Configuration:
#   http://www.ibb.net/~anne/keyboard.html
# The Linux Documentation Project: http://www.tldp.org/
# The Linux Cookbook: http://www.tldp.org/LDP/linuxcookbook/html/
# Greg's Wiki http://mywiki.wooledge.org/

if [ -z "${PROFILEREAD+true}" ] ; then

  # Setup some default paths. Note that this order will allow user installed
  # software to override 'system' software.
  # Modifying these default path settings can be done in different ways.
  # To learn more about startup files, refer to your shell's man page.

  : ${ORIGINAL_PATH=${PATH}}
  # setting CYGWIN_USEWINPATH non-empty in the system variables
  # assumes that you've already set up PATH so that Cygwin works
  # correctly -- no further alteration is done
  if [ ${CYGWIN_USEWINPATH-pathprepend} = "pathprepend" ] ; then
    # setting CYGWIN_NOWINPATH non-empty in the system variables
    # prevents use of the existing PATH and a clean PATH just for
    # Cygwin is set up -- you need to add any extra path components
    # you need in your personal startup files
    if [ ${CYGWIN_NOWINPATH-addwinpath} = "addwinpath" ] ; then
	PATH="/usr/local/bin:/usr/bin${PATH:+:${PATH}}"
    else
	PATH="/usr/local/bin:/usr/bin"
    fi
  fi

# see https://cygwin.com/ml/cygwin/2014-05/msg00352.html
# MANPATH="/usr/local/man:/usr/share/man:/usr/man${MANPATH:+:${MANPATH}}"
  INFOPATH="/usr/local/info:/usr/share/info:/usr/info${INFOPATH:+:${INFOPATH}}"

  # Set the user id
  USER="$(/usr/bin/id -un)"

  # TMP and TEMP as defined in the Windows environment
  # can have unexpected consequences for cygwin apps, so we define
  # our own to match GNU/Linux behaviour.
  unset TMP TEMP
  TMP="/tmp"
  TEMP="/tmp"

  # Define default printer
  p='/proc/registry/HKEY_CURRENT_USER/Software/Microsoft/Windows NT/CurrentVersion/Windows/Device'
  if [ -e "${p}" ] ; then
    read -r PRINTER < "${p}"
    PRINTER=${PRINTER%%,*}
  fi
  unset p

  # Default to removing the write permission for group and other
  #  (files normally created with mode 777 become 755; files created with
  #  mode 666 become 644)
  umask 022

  # Here is how HOME is set, in order of priority, when starting from Windows
  #  1) From existing HOME in the Windows environment, translated to a Posix path
  #  2) from /etc/passwd, if there is an entry with a non empty directory field
  #  3) from HOMEDRIVE/HOMEPATH
  #  4) / (root)
  # If the home directory doesn't exist, create it.
  if [ ! -d "${HOME}" ]; then
    if mkdir -p "${HOME}"; then
      echo "Copying skeleton files."
      echo "These files are for the users to personalise their cygwin experience."
      echo
      echo "They will never be overwritten nor automatically updated."
      echo
      if ! cd /etc/skel; then
        echo "WARNING: Failed attempt to cd into /etc/skel!"
      else
        /usr/bin/find . -type f |
	  while read f; do
            fDest=${f#.}
            if [ ! -e "${HOME}${fDest}" -a ! -L "${HOME}${fDest}" ]; then
              /usr/bin/install -D -p -v "${f}" "${HOME}/${fDest}"
            fi
          done
        unset fDest
      fi
    else
      echo "${HOME} could not be created."
      { [ -d "${TEMP}" ] && HOME="${TEMP}"; } ||
      { [ -d "${TMP}" ] && HOME="${TMP}"; } ||
      { [ -d /tmp ] && HOME=/tmp; } ||
      HOME=/
      echo "Setting HOME to ${HOME}."
    fi
  fi

  readonly PROFILEREAD=true
fi

# Make sure we start in home unless invoked by CHERE
if [ ! -z "${CHERE_INVOKING}" ]; then
  unset CHERE_INVOKING
else
  cd "${HOME}" || echo "WARNING: Failed attempt to cd into ${HOME}!"
fi

# Shell dependent settings
profile_d ()
{
  _LC_ALL_SET_="${LC_ALL+set}"
  _LC_SAVE_="${LC_ALL-null}"
  LC_ALL=C
  if [ "${_LC_SAVE_}" = "null" ]; then
    for file in /etc/profile.d/*.$1; do
      [ -e "${file}" ] && . "${file}"
    done
    unset LC_ALL
  else
    for file in /etc/profile.d/*.$1; do
      [ -e "${file}" ] && LC_ALL="${_LC_SAVE_}" . "${file}"
    done
    LC_ALL="${_LC_SAVE_}"
  fi
  unset file
  unset _LC_ALL_SET_
  unset _LC_SAVE_
}

HOSTNAME="$(/usr/bin/hostname)"
profile_d sh
if [ ! "x${BASH_VERSION}" = "x"  ]; then
  [ -f "/etc/bash.bashrc" ] && . "/etc/bash.bashrc"
elif [ ! "x${KSH_VERSION}" = "x" ]; then
  typeset -l HOSTNAME
  [ "${PS1-null}" = "null" ] || PS1=$(print '\033]0;${PWD}\n\033[32m${USER}@${HOSTNAME} \033[33m${PWD/${HOME}/~}\033[0m\n$ ')
elif [ ! "x${ZSH_VERSION}" = "x" ]; then
  # zsh is in shell compatibility mode here, so we probably shouldn't do this
  profile_d zsh
elif [ ! "x${POSH_VERSION}" = "x" ]; then
  : # [ "${PS1-null}" = "null" ] || PS1="$ "
else
  : # [ "${PS1-null}" = "null" ] || PS1="$ "
fi

# export PROFILEREAD PATH ORIGINAL_PATH MANPATH INFOPATH USER TMP TEMP PRINTER HOSTNAME PS1 SHELL tmp temp
export PROFILEREAD PATH ORIGINAL_PATH INFOPATH USER TMP TEMP PRINTER HOSTNAME PS1 SHELL
```

**Reply by RT-Tap** [reply]
So actually I just notice that in PowerShell I get a `NO CONFIG` message in that field, without the error message above the prompt (so just the theme seems to be broken without an error message above the prompt like in Cygwin bash).  But only for this theme (maybe others as well?).  Not sure what that field is/displays exactly because it isn't included in other themes but there is obviously some issue system wide.

Anyway this is what is in `C:\Users\ME\Documents\WindowsPowerShell\profile.ps1`:
```
# !! Contents within this block are managed by 'conda init' !!
(& "C:\Users\Artyt\anaconda3\Scripts\conda.exe" "shell.powershell" "hook") | Out-String | Invoke-Expression
oh-my-posh --init --shell pwsh --config ~/blue-owl.omp.json | Invoke-Expression
```

**Reply by JanDeDobbeleer** [reply]
Are you sure you have those files in your `$HOME` directory? As I have a working git bash shell on my machine with the following line in my `.bashrc`: `eval "$(oh-my-posh --init --shell bash --config ~/.jandedobbeleer.omp.json)"`

**Reply by RT-Tap** [reply]
Yea. I'm pretty sure.  
So powershell doesn't have this issue, not sure wtf I did or what happened but the issue is gone.
Now for Cygwin (using bash), I did `echo $HOME` it printed `/home/Artyt` and that's where my `.bashrc` file is with the `eval "$(oh-my-posh --init --shell bash --config ~/.theme.omp.json)"` line in it.  Also in that folder I have a  symlink named `.poshthemes` that links to the folder where the themes for my Windows install are. 

If I start bash from `/bin/bash` the `eval` line get's evaluated because oh-my-posh starts, however with the error : `Oh My Posh Error: CreateFile /home/Artyt/iterm2.omp.json: The system cannot find the path specified.` and the `INVALID COFIG PATH` in the actual prompt.  I tried replacing the theme name and running `. ~/.bashrc` but nothing happens I just get the same Error message.
This leads me to ask two questions:
Why is oh-my-posh trying to create the file `~/theme.omp.json`? Should it not be `~/.poshthemes/theme.omp.json`?
What type of config file are they talking about here?  A config file for oh-my-posh or a bash/shell profile?

Also,  I also use `Eugeney/Tabby` (used to be named Terminus - If your unfamiliar it's like fluent terminal or just one of those terminal emulators where you can run multiple shells and/or SSH connections in a tabbed window) and I could have sworn that when I first installed oh-my-posh it worked/showed up correctly in Tabby (for Cygwin that is, no problems with powershell), but now it's just the standard prompt.  It shouldn't be though because if i run `/bin/bash` it has the oh-my-posh prompt albeit one with an error.  But maybe there is something that can be said about the discrepancy.

**Reply by JanDeDobbeleer** [reply]
Did you symlink the `.poshthemes` folder or each theme individually? Otherwise you need to also include the folder in the path OR symlinks simply don't work for this use-case.

**Comment by mreduar** [comment]
Hi, I am also having the same problem with Cygwin.
Is there any solution for this?
![image](https://user-images.githubusercontent.com/22550119/154317296-312549ed-d897-4542-9112-853270453650.png)
I also try with `eval "$(oh-my-posh --init --shell bash --config ~/craver.omp.json)"`  and it is the same

I use Git Bash and everything works perfect there, but I would like to move to Cygwin, however I get this error.

**Reply by JanDeDobbeleer** [reply]
You should use Windows paths as you're still on Windows and golang needs to use the Windows API's. `~` won't work as it resolves to the same path. Find the actual location and use that.

**Reply by mreduar** [reply]
Thanks
`eval "$(oh-my-posh --init --shell bash --config C:/Users/<username>/cygwin64/home/<username>/<config.json>)"`  works

**Reply by rohit-kumar-j** [reply]
Thanks @mreduar , I did a similar thing, but cloned the git repo with all the themes in my `.config` folder and ran it like this... 

```bash
eval "$(oh-my-posh --init --shell bash --config C:/cygwin64/home/jsk71/.config/oh-my-posh/themes/jandedobbeleer.omp.json)"
```
it worked!

**Comment by GameLord2011** [comment]
This is probably a very dumb question, but how does one install it in the first place, can I just insert the path to the Windows executable, as cygwin is just a POSIX compatibility layer for Windows?

**Reply by JanDeDobbeleer** [reply]
Not really as the Windows path doesn't exist. All it needs is to have the executable in your `PATH` which will simply make it available.

**Reply by GameLord2011** [reply]
It seems to use the Windows PATH,
```bash
*****@************** ~
$ echo $PATH
/usr/local/bin:/cygdrive/c/Python313/Scripts:/cygdrive/c/Python313:/cygdrive/c/Program Files (x86)/Common Files/Oracle/Java/java8path:/cygdrive/c/Program Files (x86)/Common Files/Oracle/Java/javapath:/cygdrive/c/Program Files/Microsoft/jdk-21.0.7.6-hotspot/bin:/cygdrive/c/Python312/Scripts:/cygdrive/c/Python312:/cygdrive/c/windows/system32:/cygdrive/c/windows:/cygdrive/c/windows/System32/Wbem:/cygdrive/c/windows/System32/WindowsPowerShell/v1.0:/cygdrive/c/windows/System32/OpenSSH:/cygdrive/c/Program Files/HP/HP One Agent:/cygdrive/c/Program Files/dotnet:/cygdrive/c/Program Files/Microsoft SQL Server/150/Tools/Binn:/cygdrive/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/cygdrive/c/Program Files/Git/cmd:/cygdrive/c/ProgramData/chocolatey/bin:/cygdrive/c/WINDOWS/system32:/cygdrive/c/WINDOWS:/cygdrive/c/WINDOWS/System32/Wbem:/cygdrive/c/WINDOWS/System32/WindowsPowerShell/v1.0:/cygdrive/c/WINDOWS/System32/OpenSSH:/cygdrive/c/Program Files/Git LFS:/cygdrive/c/Program Files/GitHub CLI:/cygdrive/c/Users/*****/AppData/Roaming/Python/Python312/Scripts:/cygdrive/c/Program Files/nodejs:/cygdrive/c/Program Files/PowerShell/7:/usr/bin:/cygdrive/c/Users/*****/AppData/Local/pnpm:/cygdrive/c/Users/*****/AppData/Local/Microsoft/WindowsApps:/cygdrive/c/Users/*****/AppData/Local/Programs/oh-my-posh/bin:/cygdrive/c/Users/*****/AppData/Local/Programs/Microsoft VS Code/bin:/cygdrive/c/Users/*****/.dotnet/tools:/cygdrive/c/Users/*****/OneDrive/Documents/PowerShell/Scripts:/cygdrive/c/Users/*****/AppData/Local/Microsoft/WindowsApps:/cygdrive/c/Users/*****/.dotnet/tools:/cygdrive/c/Users/*****/AppData/Local/Microsoft/WinGet/Links:/cygdrive/c/Users/*****/AppData/Roaming/npm
```

(I replaced my hostname w/ the *'s)

**Reply by JanDeDobbeleer** [reply]
If you want to use oh-my-posh and not the full path (which would need to be cygwin and not Windows path, you need to make sure the oh-my-posh.exe file is available in one of those folders. Now, the installer will do that for you (restart the terminal after install to reload the `PATH`).
